Add low-level APIs for running Emscripten programs in xterm.js

mame opened the pull request:
Background: https://github.com/emscripten-core/emscripten/issues/16352

The main purpose of this changeset is to make (some kinds of) `ioctl` syscall hookable:

* `ioctl(TCGETS)` calls `TTY.default_tty_ops.ioctl_tcgets`
* `ioctl(TCSETS)`, `ioctl(TCSETSW)` and `ioctl(TCSETSF)` call `TTY.default_tty_ops.ioctl_tcsets`
* `ioctl(TCIOCGWINSZ)` calls `TTY.default_tty_ops.ioctl_tiocgwinsz`

So we can hook the events by defining `TTY.default_tty_ops.ioctl_tcgets`, etc.

These hooks serve as low-level APIs to run an Emscripten CUI program on xterm.js.
Currently, [xterm-pty](https://github.com/mame/xterm-pty) uses [very dirty hack](https://github.com/mame/xterm-pty/blob/00aae1b24b8a1bb1650643b7e1b9c1aed1f6e7cf/src/emscriptenHack.ts#L180-L212) to hook the syscall (by redefining `wasmImports`).
This hack is very fragile and will stop working if the Emscripten runtime changes (for example, `asmLibraryArg` was renamed to `wasmImports`).

The proposed ioctl hooks allow (relatively) clean integration between xterm-pty and the Emscripten runtime.
What do you think?

sbc100 commented on src/library_tty.js:
```diff
@@ -161,6 +161,27 @@ mergeInto(LibraryManager.library, {
+      ioctl_tcgets: function(tty) {
+        // typical setting
+        return {
+          c_iflag: 0x6500, // ICRNL | IXON | IMAXBEL | IUTF8
```
Can you use `{{{ cDefs.ICRNL | cDefs.IXON | cDefs.IMAXBEL | cDefs.IUTF8 }}}` here instead of the hex number?  Or is that harder to read?

mame replied:
Done at 1145e3975d190da10b1ee2c663f8da3ee8f535eb

sbc100 commented on src/library_syscall.js:
```diff
@@ -544,7 +588,13 @@ var SyscallsLibrary = {
+        var timeoutInMillis = -1;
+        if (timeout) {
+          var timeoutLow = (readfds ? {{{ makeGetValue('timeout', 0, 'i32') }}} : 0),
+              timeoutHigh = (readfds ? {{{ makeGetValue('timeout', 4, 'i32') }}} : 0);
```
I think you can use ` var timeout = {{{ makeGetValue('timeout', 4, 'i53') }}}` instead of reading the hi/low pair yourself.

mame replied:
Oh sorry, this is intentional here. The code reads the following two fields of `struct timeval`.

```
struct timeval {
  long tv_sec;   // seconds
  long tv_usec; // microseconds
};
```

The names `timeoutLow` and `timeoutHigh` were bad. I'll change them as `timeoutSeconds` and `timeoutMicroseconds`.

mame replied:
I changed the variable names d8efb50a537df5d1179c6af79b1013a9b2a6185a

sbc100 replied:
In that case I wonder if its worth using `timeval.tv_sec` and `timeval.tv_usec` here rather than hardcoding 0 and 4?  Maybe not worth it?

mame replied:
You are absolutely right. Why didn't I notice it? Done at 8fe4bb834a4936ce2f73bd1ad692eea95ddad070

sbc100 reviewed (commented):
Looks great! 

Is there some way we can test this here in the emscripten repo to avoid regressions?

mame commented on src/library_syscall.js:
```diff
@@ -209,21 +209,55 @@ var SyscallsLibrary = {
       case {{{ cDefs.TCSETSF }}}: {
         if (!stream.tty) return -{{{ cDefs.ENOTTY }}};
+        if (stream.tty.ops.ioctl_tcsets) {
+          var optional_actions = op == {{{ cDefs.TCSETS }}} ? "" : op == {{{ cDefs.TCSETSW }}} ? "W" : "F";
```
It is an argument to distinguish the event types: `""` stands for `TCSETS`, `"W"` does for `TCSETSW`, and `"F"` does for `TCSETSF`.

I am open to change the values. Do you like to pass the numbers like `cDefs.TCSETS` or `cDefs.TCSETSW`? Or pass a string like `"TCSETS"` or `"TCSETSW"`?

sbc100 replied:
Maybe just stick with `cDefs.TCSETS` since they already exist?

sbc100 replied:
Maybe just stick with `cDefs.TCSETS` since they already exist?

mame replied:
I did so, thanks! 16fff9d5175bca1741aa5fca35e06518fb761e71

mame replied:
Just FYI: The reason why I did something a little different is that the constants such as `cDefs.TCSETS` cannot be referred from xterm-pty which is outside of Emscripten. But, the values of the constants probably won't change easily, so it's not a big deal. I think we should aim to incorporate xterm-pty into Emscripten in the end.

sbc100 replied:
Right that makes sense.

sbc100 replied:
> Just FYI: The reason why I did something a little different is that the constants such as `cDefs.TCSETS` cannot be referred from xterm-pty which is outside of Emscripten. But, the values of the constants probably won't change easily, so it's not a big deal. I think we should aim to incorporate xterm-pty into Emscripten in the end.

BYW if you write the glue code between xterm.js and emscripten as a JS library file you can access all the macros.  e.g. `--js-library=library_xterm.js` .. can code than interpret the emscripten requests and make calls into the external xterm.js library.

mame replied:
Yes, I would like to challenge that next.

However, I think it may be difficult to get everything working nice with just `--js-library=library_xterm.js`.

This is because `getchar` function needs to synchronously wait for user input. The default implementation of `getchar` waits synchornously by using `prompt` function, but we can't do the same for xterm.js. So, xterm-pty requires to run the Emscripten runtime in a Web Worker, and waits for user input via SharedArrayBuffer.

Therefore, `emcc` needs to generate two JS files, one for the main thread and one for the Web Worker. I wonder if this is beyond the scope of the `--js-library=library_xterm.js` option.

sbc100 replied:
We have an option called PROXY_TO_THREAD that puts the main application code in worker.   When `getchar` is called from a worker it should be able to block..   We also have `-sASYNCIFY` and `-sASYNCIFY=2` (A.K.A JSPI) which allow functions to suspend/block, even when running on the main thread.

mame replied:
Great to know PROXY_TO_THREAD, thanks! I will read the code.

> We also have `-sASYNCIFY` and `-sASYNCIFY=2` (A.K.A JSPI) which allow functions to suspend/block, even when running on the main thread.

I'm not sure about this. As long as `getchar` is not an async function, I thought it could not wait synchronously in the main thread. Maybe I need to learn JSPI.

sbc100 replied:
`-sASYNCIFY` and `-sASYNCIFY=2` allow the wasm module to suspend its computation and resume later.  For example, after some async work is done.  From the POV of the code being suspended it appears like a blocking call.

mame commented:
> Looks great!

Thank you!

> Is there some way we can test this here in the emscripten repo to avoid regressions?

Sure, I guess the last additional commit should fix the issue.

(It may be good to write "run `. /tools/gen_struct_info.py --wasm64`" in [this section of the documentation](https://emscripten.org/docs/contributing/developers_guide.html#working-with-c-structs-and-defines).)

mame commented:
@sbc100 Thank you, I think I've addressed all your review comments and the CI has passed. Do you want me to tidy up the commits by rebase? Please feel free to let me know.

sbc100 commented:
In terms of testing, perhaps you could modify `test/other/test_ioctl.c`?

mame commented:
> In terms of testing, perhaps you could modify `test/other/test_ioctl.c`?

Thanks! I have added `test/other/test_ioctl_termios.c`.

appden commented:
Excited to see this land. Looks like this was ready to auto-merge then the force push cancelled it.
